File: inbound/src/roles.ts
// Role-address membership for BOUND WEBMAIL SESSIONS (#425, the webmail half of the
// #404 ruling of 2026-07-25).
//
// A role address (abuse@, security@, support@) belongs to a FUNCTION, so under a
// per-viewer door it is nobody address and its mail is delivered, stored, searchable
// and visible to NO human. The IMAP door closed that with POSTERN_IMAP_VIEWER_ROLES
// (PR #423): a viewer resolves to a SET -- its own address V plus every role V is a
// member of -- and each role publishes as its OWN folder, never merged into INBOX.
// Webmail scopes to the single bound session identity, so without this module the two
// human doors disagree about what one person may see.
//
// FORMAT (identical to the door, deliberately): role=member+member,role2=member --
// commas separate roles, + separates members, and BOTH sides are full mail addresses.
// Everything is lower-cased and whitespace around any token is tolerated.
//
// WHY A SECOND VAR, AND WHAT IT COSTS. The door parses its own env at startup and
// must keep doing so: making it fetch membership from the worker would put a network
// dependency in front of a fail-closed startup, so one flake would read as "nobody is
// on any queue". The repo precedent is the other direction and it works -- the
// AUTH_THROTTLE_* knobs are ONE vocabulary configured verbatim on two doors. The cost
// is operator DRIFT: POSTERN_VIEWER_ROLES here and POSTERN_IMAP_VIEWER_ROLES on the
// door can disagree, which is the very divergence #425 exists to close. Three things
// bound it: the refusal set below is a port of the door _parse_viewer_roles, so a
// config the door refuses to START on is refused here too (one config is legal or
// illegal identically on both doors); the mirroring rule and deploy ordering are
// documented (docs/CONTRACT.md, webmail/README.md, imap/README.md); and GET /api/roles
// exposes the parsed map to an operator token so the two can be DIFFED instead of
// assumed. Single-source promotion is tracked as a follow-up, not done here.
//
// FAIL-CLOSED, AND WHY IT DIFFERS FROM FILE_ALSO_UNDER. ANY malformed or ambiguous
// entry drops the WHOLE map (not just that entry): a silently dropped member reads
// exactly like "that person is not on the queue", which is the #404 failure mode
// again. FILE_ALSO_UNDER skips-and-logs per entry because it sits on the DELIVERY
// path, where refusing costs MAIL; this sits on the READ path, where refusing costs
// ACCESS only. Losing access to a queue is loud and reversible by fixing the config;
// half a membership map is silent and is not.

/** A parse refusal. Never surfaced to a caller: it drops the map and is logged. */
class RoleConfigError extends Error {}

// Parse cache. A Worker has no startup phase to fail at, so the map is parsed on
// demand and memoized on the RAW STRING: a config change (new deploy, new isolate)
// re-parses, and a hot isolate never re-parses the same text.
let cachedRaw: string | null = null;
let cachedMap: ReadonlyMap<string, readonly string[]> = new Map();

const EMPTY: ReadonlyMap<string, readonly string[]> = new Map();

/** Parse POSTERN_VIEWER_ROLES into {role: [member, ...]}, or THROW (#425).
 *
 *  Exported for the suite: the caller-facing entry point (roleMap) turns every throw
 *  into an empty map, so the refusal set is only assertable directly. The refusals are
 *  a port of imap/posternimap/config.py _parse_viewer_roles, entry for entry:
 *  no "=", an empty side, a missing "@" on either side, a role listed twice, a role
 *  that lists itself, an address used as BOTH a role and a member, a role with no
 *  members, and two roles whose local parts collide (which is a DOOR folder-naming
 *  constraint, kept here purely so both doors accept exactly the same configs). */
export function parseViewerRoles(raw: string | undefined | null): Map<string, string[]> {
  const out = new Map<string, string[]>();
  if (!raw || !raw.trim()) return out;
  const localParts = new Map<string, string>();
  for (const chunk of raw.split(",")) {
    const entry = chunk.trim();
    if (!entry) continue;
    const eq = entry.indexOf("=");
    if (eq < 0) {
      throw new RoleConfigError(`entries must be role=member+member (got ${entry})`);
    }
    const role = entry.slice(0, eq).trim().toLowerCase();
    if (!role || !role.includes("@")) {
      throw new RoleConfigError(
        `entry ${entry} needs a full role address on the left (e.g. abuse@example.org=ada@example.org)`,
      );
    }
    if (out.has(role)) {
      throw new RoleConfigError(
        `lists the role ${role} twice; put every member of a role in ONE entry (role=a@x+b@x)`,
      );
    }
    const local = role.slice(0, role.indexOf("@"));
    const clash = localParts.get(local);
    if (clash) {
      throw new RoleConfigError(
        `roles ${clash} and ${role} share the local part ${local}, so both would publish as one ` +
          "folder; give them distinct local parts",
      );
    }
    const members: string[] = [];
    for (const rawMember of entry.slice(eq + 1).split("+")) {
      const member = rawMember.trim().toLowerCase();
      if (!member || !member.includes("@")) {
        throw new RoleConfigError(
          `entry ${entry} needs full member addresses on the right, separated by + ` +
            "(e.g. abuse@x.org=ada@x.org+ben@x.org)",
        );
      }
      if (member === role) {
        throw new RoleConfigError(`role ${role} lists itself as a member`);
      }
      if (!members.includes(member)) members.push(member);
    }
    if (members.length === 0) {
      throw new RoleConfigError(
        `role ${role} has no members; remove the role or give it at least one member address`,
      );
    }
    localParts.set(local, role);
    out.set(role, members);
  }
  const everyMember = new Set<string>();
  for (const members of out.values()) for (const m of members) everyMember.add(m);
  const both = [...out.keys()].filter((role) => everyMember.has(role)).sort();
  if (both.length) {
    throw new RoleConfigError(
      `uses ${both.join(", ")} as BOTH a role address and a member address; an address is ` +
        "either a queue or a person, never both",
    );
  }
  return out;
}

/** The parsed role map, or an EMPTY map if the config is unusable (#425).
 *
 *  Unset, empty, or unparseable all resolve to no roles at all: no role folder is
 *  enumerated, no role read is possible, and every surface behaves exactly as it did
 *  before this feature. The log names the offending ENTRY (operator config text, an
 *  address at worst, never a credential) rather than dumping the whole raw var, so the
 *  line stays greppable. */
export function roleMap(env: Env): ReadonlyMap<string, readonly string[]> {
  const raw = env.POSTERN_VIEWER_ROLES ?? "";
  if (raw === cachedRaw) return cachedMap;
  let parsed: ReadonlyMap<string, readonly string[]>;
  try {
    parsed = parseViewerRoles(raw);
  } catch (err) {
    // Fail CLOSED on the whole map (see the header note): a partially applied
    // membership map is indistinguishable from "that person is not on the queue".
    console.error(
      "POSTERN_VIEWER_ROLES: refusing the ENTIRE map, no role queue is served:",
      err instanceof Error ? err.message : String(err),
    );
    parsed = EMPTY;
  }
  cachedRaw = raw;
  cachedMap = parsed;
  return parsed;
}

/** The role addresses this viewer may read, in config order (#425).
 *
 *  A viewer of undefined/empty yields NOTHING: membership is unanswerable without V,
 *  and the fail-closed answer is no roles, never a guess. Mirrors the door
 *  role_folders_for, which makes the same call for the same reason. */
export function rolesForViewer(env: Env, viewer: string | undefined | null): string[] {
  const v = viewer?.trim().toLowerCase();
  if (!v) return [];
  const out: string[] = [];
  for (const [role, members] of roleMap(env)) {
    if (members.includes(v)) out.push(role);
  }
  return out;
}

/** Reset the parse cache. Test-only seam: suites drive many env shapes through one
 *  module instance, and a raw-string cache would otherwise leak the first one. */
export function resetRoleCache(): void {
  cachedRaw = null;
  cachedMap = EMPTY;
}

File: imap/posternimap/roles.py
"""Role-address membership, read from the Worker (#438).

SINGLE SOURCE. Membership used to be configured TWICE: POSTERN_VIEWER_ROLES on the
Worker (the webmail half, #425) and POSTERN_IMAP_VIEWER_ROLES here (#404), the same
syntax with the same refusal set, kept equal by an operator. Two configurations of one
fact drift, and the drift IS the divergence #425 exists to close: whichever side is
broader shows a queue to someone the other side does not. The Worker map is now the only
source, and this module reads it.

WHY A FETCH IS SAFE HERE, WHICH IS NOT WHAT #425 CONCLUDED. #425 refused a fetch because
it would put a network dependency in front of a FAIL-CLOSED STARTUP: one flake at boot
would read as "nobody is on any queue". That reasoning was about STARTUP, and nothing
here runs there. The door is a PROXY, not a store: LIST, SELECT, FETCH and /api/folders
are all Worker calls, so a Worker this door cannot reach is already a mailbox with no
mail in it. Reading membership over the same connection adds no failure mode the session
does not already have. The fetch happens per session on the first LIST or SELECT, both of
which Twisted runs through the thread pool (threaded.ThreadedAccount), so it never blocks
the reactor -- the synchronous account accessors read a map that is already in memory.

FAIL-CLOSED, UNCHANGED. A fetch that fails -- unreachable, timed out, 401/403 from a
mis-provisioned token, 404 from a Worker older than #438 -- yields NO roles for that
session, loudly, and is retried on the next login. There is deliberately NO on-disk
last-known-good: the door ships as a stateless container, and a cache that outlived the
process would keep a REVOKED member reading a queue after the operator removed them.
Staleness is bounded by the TTL, and only in the permissive direction.

WHERE THE REFUSAL SET LIVES NOW. The Worker parses POSTERN_VIEWER_ROLES and drops the
WHOLE map on any malformed or ambiguous entry (inbound/src/roles.ts): an unusable config
projects as an EMPTY map, so this door inherits that refusal by construction rather than
re-implementing it and drifting from it, which is the whole point of #438. What is
validated HERE is only what the Worker cannot own: the STRUCTURE of the response, because
a door that built folders out of arbitrary JSON would be trusting a parser it does not
have; and the one DOOR invariant, that two roles must not share a local part, because
both would then publish as one folder name. Either kind of violation drops the WHOLE map,
for the Worker reason exactly: half a membership map is indistinguishable from having
been taken off the queue.
"""

from __future__ import annotations

import logging
import threading
import time
from typing import Any, Dict, Mapping, Optional, Tuple

from .client import PosternClient, PosternError
from .config import Config, role_folder_name

_log = logging.getLogger(__name__)

# Process-wide cache: (expiry on the monotonic clock, map). Shared by every session
# because the map is ESTATE config and not a per-login fact, so the door pays one fetch
# per TTL rather than one per login. Only a SUCCESSFUL parse is ever stored: a failed
# fetch or a refused payload caches nothing, so the next login retries instead of
# inheriting a failure.
_LOCK = threading.Lock()
_CACHE: Optional[Tuple[float, Mapping[str, tuple]]] = None


class RolePayloadError(ValueError):
    """The Worker answered with something this door will not build folders from."""


def parse_role_payload(payload: Any) -> Dict[str, tuple]:
    """Validate the GET /api/imap/roles body into {role: (member, ...)}, or raise.

    STRUCTURE plus one door invariant, and nothing else: the semantic refusals (a role
    that lists itself, an address used as both a queue and a person, a role with no
    members declared twice, ...) are the WORKER parser refusals, and a config that trips
    them never reaches this door as a map at all -- it arrives as the empty projection.
    Re-deriving them here would recreate the second parser #438 exists to delete.
    """
    if not isinstance(payload, dict):
        raise RolePayloadError("the response is not a JSON object")
    roles = payload.get("roles")
    if not isinstance(roles, list):
        raise RolePayloadError("the response carries no roles list")
    out: Dict[str, tuple] = {}
    localparts: Dict[str, str] = {}
    for entry in roles:
        if not isinstance(entry, dict):
            raise RolePayloadError("a roles entry is not an object")
        address = entry.get("address")
        if not isinstance(address, str):
            raise RolePayloadError("a roles entry carries no address string")
        role = address.strip().lower()
        if not role or "@" not in role:
            raise RolePayloadError(
                "role address %r is not a full mail address" % address
            )
        if role in out:
            raise RolePayloadError("the role %r is listed twice" % role)
        local = role.split("@", 1)[0]
        if local in localparts:
            raise RolePayloadError(
                "roles %r and %r share the local part %r, so both would publish as the "
                "folder %s; give them distinct local parts"
                % (localparts[local], role, local, role_folder_name(role))
            )
        raw_members = entry.get("members")
        if not isinstance(raw_members, list):
            raise RolePayloadError("the role %r carries no members list" % role)
        members: list = []
        for raw in raw_members:
            if not isinstance(raw, str):
                raise RolePayloadError("the role %r has a non-string member" % role)
            member = raw.strip().lower()
            if not member or "@" not in member:
                raise RolePayloadError(
                    "the role %r has the member %r, which is not a full mail address"
                    % (role, raw)
                )
            if member not in members:
                members.append(member)
        if not members:
            raise RolePayloadError("the role %r has no members" % role)
        localparts[local] = role
        out[role] = tuple(members)
    return out


def viewer_roles(cfg: Config, client: Optional[PosternClient]) -> Mapping[str, tuple]:
    """The role map this door should act on: fetched, cached, or EMPTY.

    Empty is the fail-closed answer everywhere it appears, and every path that produces
    it says so out loud, because a queue that quietly stops being visible is the #404
    failure mode this feature exists to end.
    """
    if cfg.viewer_mode != "per_account":
        # Estate mode has no viewer address to check membership against, so a role folder
        # is unreachable by construction (role_folders_for refuses a viewer of None) and
        # there is nothing to fetch a map FOR. This is not the silent no-op the old
        # startup error guarded against: the map still applies on the Worker, which is
        # where it is configured and where webmail reads it.
        return {}
    if client is None:
        _log.warning(
            "postern-imap: POSTERN_IMAP_VIEWER_MODE=per_account without "
            "POSTERN_API_TOKEN_IMAP. Role membership is read from the Worker over the "
            "imap-scoped seam since #438, so NO role queue can be served until that "
            "per-function token is provisioned."
        )
        return {}

    global _CACHE
    with _LOCK:
        cached = _CACHE
        if cached is not None:
            if cached[0] > time.monotonic():
                return cached[1]
            # EXPIRED, so drop it before the refetch: a refetch that fails must not fall
            # back onto a stale membership, or a member the operator removed keeps the
            # queue. The TTL is the revocation bound and this is what enforces it.
            _CACHE = None

    try:
        parsed = parse_role_payload(client.get_roles())
    except PosternError as exc:
        _log.warning(
            "postern-imap: cannot read role membership from the Worker (%s); serving NO "
            "role queue this session, retrying on the next login. A 404 means the Worker "
            "predates #438 (deploy the Worker first); a 403 means the imap-scoped token "
            "is not the one the Worker gates on.",
            exc,
        )
        return {}
    except RolePayloadError as exc:
        _log.warning(
            "postern-imap: refusing the ENTIRE role map the Worker returned (%s). Half a "
            "membership map reads exactly like being taken off the queue, so no role "
            "queue is served at all.",
            exc,
        )
        return {}

    with _LOCK:
        _CACHE = (time.monotonic() + cfg.roles_ttl, parsed)
    return parsed


def reset_cache() -> None:
    """Drop the cached map.

    Test seam, and the honest kind: a suite drives many Worker answers through one
    module instance, and a process-wide cache would otherwise leak the first one into
    every case after it (the same reason inbound/src/roles.ts exports resetRoleCache).
    """
    global _CACHE
    with _LOCK:
        _CACHE = None
